### ISSUE TITLE
Transaction lost after broadcast

### ISSUE THREAD

**Issue by abitmore** [bitshares/bitshares-core#2376]
**Bug Description**

It was reported in https://github.com/bitshares/bitshares-core/issues/490 that transactions can sometimes get lost without being propagated. It's believed that the issue occurs more frequently when the node is busy.

BTW there were efforts to get around the issue in Steem (see https://github.com/steemit/steem/commit/0fb40572c74791f85a0152c5bceb68ec9331e6c3) and an attempt to port it to BitShares (see https://github.com/bitshares/bitshares-core/pull/703).

I think it is due to the use of `std::unordered_set` which is not thread-safe.
```
        std::unordered_set<item_id> inventory_to_advertise;
        inventory_to_advertise.swap(_new_inventory);
```

(Update: there were some logs here which are removed later because they were for another issue(#2378).)

**Impacts**
Describe which portion(s) of BitShares Core may be impacted by this bug. Please tick at least one box.
- [ ] API (the application programming interface)
- [ ] Build (the build process or something prior to compiled code)
- [ ] CLI (the command line wallet)
- [ ] Deployment (the deployment process after building such as Docker, Travis, etc.)
- [ ] DEX (the Decentralized EXchange, market engine, etc.)
- [x] P2P (the peer-to-peer network for transaction/block propagation)
- [ ] Performance (system or user efficiency, etc.)
- [ ] Protocol (the blockchain logic, consensus, validation, etc.)
- [ ] Security (the security of system or user data, etc.)
- [ ] UX (the User Experience)
- [ ] Other (please add below)

## CORE TEAM TASK LIST
- [ ] Evaluate / Prioritize Bug Report
- [ ] Refine User Stories / Requirements
- [ ] Define Test Cases
- [ ] Design / Develop Solution
- [ ] Perform QA/Testing
- [ ] Update Documentation

**Comment by abitmore** [comment]
Actually, in the Github Action logs, the value of `peer->peer_needs_sync_items_from_us` is different between 2 runs. It was another issue. I'll fix that issue with a pull request but keep this issue open so far.

**Comment by abitmore** [comment]
Should be fixed by #2383.
